### PR TITLE
[HW] Add `hw.triggered` op

### PR DESCRIPTION
**Pull request by mortbopet**
Here's a stab at #5574.
I think the main issue is to decide where to place the lowering. AFAICT, the closest we have to a "hw to sv" conversion pass is `hw-legalize-modules` (which really should be named to `sv-legalize-modules` seeing as it's lowering away HW constructs in favor of SV constructs). I chose to introduce a new pass - `lower-hw-to-sv` which could serve as a clear home for these kinds of lowerings (maybe some things in `HWLegalizeModules` could be moved to here?).

Lastly, just went with a simple assembly format. If people would like to see something more fancy, e.g. an initializer-list style format with an elided entry block, let me know:
> `hw.triggered posedge %t (%a0 : i32 = %in) { ...`

### CODE REVIEW

File: lib/Dialect/HW/HWOps.cpp [hunk 1]
```diff
@@ -3202,6 +3202,23 @@ ParseResult HierPathOp::parse(OpAsmParser &parser, OperationState &result) {
...
+void TriggeredOp::build(OpBuilder &builder, OperationState &odsState,
+                        EventControlAttr event, Value trigger,
+                        ValueRange inputs) {
+  odsState.addOperands(trigger);
+  odsState.addOperands(inputs);
+  odsState.addAttribute(getEventAttrName(odsState.name), event);
+  auto *r = odsState.addRegion();
+  Block *b = new Block();
+  r->push_back(b);
+  b->addArguments(inputs.getTypes(), llvm::SmallVector<mlir::Location>(
```
Comment (teqdruid) [inline lib/Dialect/HW/HWOps.cpp:3218]: I think these locations should come from the inputs' defining ops.

**Review by teqdruid — APPROVED** [review]
Looks good to me!

Initializer-style asm format would be nice, but not necessary. It seems so nice and general; someone should add it to mlir-tblgen assemblyFormat...

File: lib/Conversion/HWToSV/HWToSV.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  switch (ec) {
+  case hw::EventControl::AtPosEdge:
+    return sv::EventControl::AtPosEdge;
+  case hw::EventControl::AtNegEdge:
+    return sv::EventControl::AtNegEdge;
+  case hw::EventControl::AtEdge:
+    return sv::EventControl::AtEdge;
+  }
+  llvm_unreachable("Unknown event control kind");
+}
```
Comment (fabianschuiki) [inline lib/Conversion/HWToSV/HWToSV.cpp:35]: Now that we have `EventControl` in HW, we might be able to replace all other dialect-local `EventControl` things with the HW flavor! I think I've seen a few dialects where we redefine that enum.

File: test/Conversion/HWToSV/test_trigger.mlir [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+hw.module @foo(%trigger : i1, %in : i32) {
+  // CHECK:       sv.always posedge %trigger {
+  // CHECK-NEXT:    "some.user"(%in) : (i32) -> ()
+  // CHECK-NEXT:  }
+  hw.triggered posedge %trigger (%in) : i32 {
+    ^bb0(%arg0 : i32):
+      "some.user" (%arg0) : (i32) -> ()
+  }
+}
```
Comment (fabianschuiki) [inline test/Conversion/HWToSV/test_trigger.mlir:11]: Cool!

**Review by fabianschuiki — APPROVED** [review]
LGTM!

I'm starting to wonder if we should have a `LowerCoreToSV` pass that combines all the Seq and HW and Comb lowerings into SV in one place. Not sure how useful that would be though :thinking:

File: lib/Conversion/HWToSV/HWToSV.cpp (hunk 2 above)
Comment (mortbopet) [reply lib/Conversion/HWToSV/HWToSV.cpp:35]: Captured in https://github.com/llvm/circt/issues/5601.

**Pull request merged**